src/main/java/org/candlepin/sync/Exporter.java

```diff
@@ -536,17 +523,34 @@ private void exportProducts(File baseDir, Consumer consumer) throws IOException
         }
     }
 
-    private void addProvidedProducts(Collection<Product> providedProducts, Map<String, Product> products) {
-        if (providedProducts == null || providedProducts.isEmpty()) {
+    /**
+     * Adds the specified product and all of its children products to the given product map, using
+     * products' ID (*not* UUID) as the key in the map. If the specified product is null or does not
+     * have a product ID, it will be silently ignored.
+     *
+     * @param product
+     *  the product to add to the map
+     *
+     * @param productMap
+     *  the map in which to collect products
+     */
+    private void collectProducts(Product product, Map<String, Product> productMap) {
+        if (product == null || product.getId() == null) {
             return;
         }
 
-        for (Product product : providedProducts) {
-            if (product != null) {
-                products.put(product.getId(), product);
-                addProvidedProducts(product.getProvidedProducts(), products);
+        productMap.put(product.getId(), product);
+
+        // Add provided products (if applicable)
+        Collection<Product> providedProducts = product.getProvidedProducts();
+        if (providedProducts != null) {
+            for (Product provided : providedProducts) {
+                this.collectProducts(provided, productMap);
```

Comment:
Thanks for the detailed explanation! I wasn't aware that the Product class was proactively checking for cycles. I feel a lot better knowing it would be super rare for an issue like that to occur naturally.